### ISSUE TITLE
Biometric protected key

### ISSUE THREAD

**Issue by stohn** [Yubico/libfido2#206]
Hi,

is there any way to specify that a key inside of FIDO2 device requires biometric verification before use?

My setup:
- enrollment of biometrics worked fine (`fido2-token -S -e ....`)
- key creation (via latest openssh) worked fine (`ssh-keygen -t ecdsa-sk -f ~/.ssh/id_ecdsa_sk_rk -O resident`)
- when logging in to a ssh server I only get asked for `Confirm user presence for key ECDSA-SK SHA256:.....`

- in openssh there is an option to specify `no-touch-required` for a key but I could not find anything related to biometrics

=> So i was checking the libfido2 API and also could not find anything.

Questions: 
* What is the enrolled biometric on fido2-token used for? 
* is there a way to enforce biometric verification for every key usage or at least for authentication after the fido2 device is plugged in?

My goal should be easy to guess: Use a FIDO2 key with fingerprint to login to ssh servers, requiring fingerprint to authorize the login (or at least use fingerprint to unlock the FIDO2 device after insertion).

Thanks,

Maik

**Comment by martelletto** [comment]
User verification (PIN or biometric) can be enforced for a given credential at the authenticator level through fido_cred_set_prot(). Use of fido_cred_set_prot() has been implemented in OpenSSH HEAD, but not yet released: `$ ssh-keygen -t ecdsa-sk -O verify-required`. At the moment, OpenSSH's understanding of FIDO2 user verification is restricted to PIN. As an experiment, you could try to comment out [this chunk](https://github.com/openssh/openssh-portable/blob/master/sshconnect2.c#L1208-L1213) in sshconnect2.c's identity_sign(), explicitly specify the key to be used by ssh (`ssh -i <key>` or add a `~/.ssh/config` host entry to that effect), and touch the authenticator with the enrolled finger. It should work. If you would like sshd to check whether user verification took place, you will need OpenSSH's HEAD sshd.

**Comment by stohn** [comment]
Thank you for the great answer.

I went down the rabbit hole and compiled openssh-portable HEAD from source on my Mac (suggestion to others, use a Linux system in a VM if you want to try this, I took a lot of time and required some manual patches to compile openssh on a recent MacOS).

With the new version I could specify the new option "verify-required" and now when I want to use the key openssh always prompts for the password.

When I enter the password I'm asked to "Confirm presence by touching" afterwards, which I can do with any finger (no fingerprint is checked).

However (THANK YOU FOR THE HINT), when I just press return (or remove the password request in source like you suggested) then openssh still asks to "Confirm presence by touching"... BUT MAGICALLY the FIDO2 token is not accepting any finger, instead it waits for the enrolled finger to be presented and if you do so... SUCCESS.

I think you nailed it with "At the moment, OpenSSH's understanding of FIDO2 user verification is restricted to PIN".

* Seems like we should report and ask for better UX of biometrics in openssh project.

Thanks again,

Maik
